service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java

```diff
@@ -164,6 +179,103 @@ private void processCredentialsMessageGetOperation(final Message<JsonObject> reg
         getCredentials(tenantId, type, authId, result -> reply(regMsg, result));
     }
 
+    private void processCredentialsMessageAddOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        if (!isValidCredentialsObject(payload)) {
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        addCredentials(tenantId, payload, result -> reply(regMsg, result));
+    }
+    
+    private void processCredentialsMessageUpdateOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        if (!isValidCredentialsObject(payload)) {
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        updateCredentials(tenantId, payload, result -> reply(regMsg, result)); 
+    }
+    
+    private void processCredentialsMessageRemoveOperation(final Message<JsonObject> regMsg, final String tenantId, final JsonObject payload) {
+        final String deviceId = payload.getString(FIELD_DEVICE_ID);
+        if (deviceId == null) {
+            log.debug("credentials remove request did not contain device-id in payload - not supported");
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+        
+        final String type = payload.getString(FIELD_TYPE);
+        if (type == null) {
+            log.debug("credentials remove request did not contain device-id in payload - not supported");
+            reply(regMsg, CredentialsResult.from(HTTP_BAD_REQUEST));
+            return;
+        }
+
+        final String authId = payload.getString(FIELD_AUTH_ID);
+        
+        removeCredentials(tenantId, deviceId, type, authId, result -> reply(regMsg, result));
+    }
+
+    boolean isValidCredentialsObject(JsonObject credentials) {
```

Comment:
why not private?

Reply:
Package private was chosen, so that it's possible to unit test the method.

Reply:
Personally I dislike having visibility changes of methods just for the purpose of unit testing.

Why not test them via the public method `processCredentialsMessage` provided?

This leaves the class clean regarding the visibility and still tests the functionality - how the class internally implements it is not very important for me - that the class behaves correctly to the callers is what counts.